### ISSUE TITLE
Powerline not working with bash - displaying ???'s for separators (powerline vim dispalys OK)

### ISSUE THREAD

**Issue by VoltairesPuppet** [powerline/powerline#1858]
Currently the bash prompt showing as colour version, but with no glyphs, i.e.:

    ?username???/???opt???path1???path2???

`vim` is working as expected; the correct glyph for the hard separators.  

**The expected glyphs *are* visible if I `cat` the `powerline.json` files with the hard separator codes.**

Originally it wasn't displaying within vim, until I installed a patched font for for putty.  Even then it still wasn't visible on the shell prompt, however I at least go the *question marks to show* by setting the locale (see towards the end) and re-copying the main config:

    /usr/lib/python2.7/site-packages/powerline/config_files/config.json 

to

     ~/.config/powerline/config.json

Calling powerline using: 

    POWERLINE_BASH_CONTINUATION=1 POWERLINE_BASH_SELECT=1 source /usr/lib/python2.7/site-packages/powerline/bindings/bash/powerline.sh

This is inside a docker instance using a centos7 image, with powerline installed using `pip` as per the setup instructions of the main powerline git repository. I did have to add the locale to the bare bones image:

    localedef -i en_GB -c -f UTF-8 en_GB.UTF-8

And I changed the locale to:

    LANG=en_GB.utf-8
    LC_CTYPE="en_GB.utf-8"
    LC_NUMERIC="en_GB.utf-8"
    LC_TIME="en_GB.utf-8"
    LC_COLLATE="en_GB.utf-8"
    LC_MONETARY="en_GB.utf-8"
    LC_MESSAGES="en_GB.utf-8"
    LC_PAPER="en_GB.utf-8"
    LC_NAME="en_GB.utf-8"
    LC_ADDRESS="en_GB.utf-8"
    LC_TELEPHONE="en_GB.utf-8"
    LC_MEASUREMENT="en_GB.utf-8"
    LC_IDENTIFICATION="en_GB.utf-8"
    LC_ALL=en_GB.utf-8

I appreciate any help to try and fix this, and any tips on how to help diagnose the issue (even code edits) is appreciated.   

Is it possible for instance to print out the current theme and what configuration files are being used? 

It seems like this could be an issue with the render engine or perhaps an incompatibility with bash (*GNU bash, version 4.2.46(2)-release (x86_64-redhat-linux-gnu)*)?

**EDIT(1):** *Using `echo $PS1 | od -c -t x1` it looks as if the question marks are in the prompt itself.   An issue with the powerline render engine perhaps?*

**Comment by VoltairesPuppet** [comment]
Technically this is cross posted on stack-overflow if you want credit for solving this there too.

https://stackoverflow.com/q/47224826/2546168

**Comment by ZyX-I** [comment]
Those question marks are to be expected if you have forced some unicode top_theme and yet run daemon in non-unicode locale. Note: do *not* check the locale you have in shell, the locale you had at the time daemon was run is what matters. Basically if you have `powerline-daemon` before setting LANG then your shell config is wrong: first, daemon is run with a wrong locale and, second, shell configuration is not a good place for locale settings at all, you should have some relevant file in `/etc` and/or terminal emulators’ own settings. The former case is there because more apps will catch those settings, the latter case because encoding used by terminal emulator must match encoding used by the shell, but shell generally is unable to configure terminal emulator, while terminal emulator can define additional environment variables (or even have dedicated locale setting) when running an application inside.

**Comment by ZyX-I** [comment]
Though if you did not run a daemon in advance (can you provide a full script for docker?) then there is another possibility: if I recall correctly some locale settings caused some version of Python report wrong encoding: the Vim version uses Vim’s &encoding which Vim deduces from locale on its own, but shell version relies on Python built-in library for this task. Quick way to rule out the daemon is `powerline-render shell left`: even if daemon was run by something I have not been told about this will not use it.

Some other quick tests to perform to debug this issue:

1. Run `powerline-render shell left` with locale set just like you have it, but with `UTF-8` encoding (all caps).
2. Run `python -c 'from powerline.lib.encoding import get_preferred_output_encoding as gpoe; print(gpoe())'`: this should report UTF-8 or else something is broken in Python (you may see the code of that function, it is pretty straightforward), check output of `locale.getlocale(locale.LC_MESSAGES)` and `locale.getdefaultlocale()` to see what.

**Comment by VoltairesPuppet** [comment]
Thanks @ZyX-I .   Looks like the daemon running in the background was the primary issue here.   I would have thought running the script again would have replaced the daemon but I guess that isn't the case.  

I had started to use `powerline-render` to try debugging the issue but it was presenting the same information; likely due to the daemon still running.   Is it one daemon per user, as I had one hanging around for root and one I think may have been from another terminal session? 

That said, I have a few things to verify if they are required / helping to resolve the issue:

* `~/.config/powerline/config.cfg`  <- user `"common"` the value pair `"default_top_theme": "powerline" ` is added
* `LC_ALL=en_GB.utf8` rather than the `en_GB.UTF-8` / `en_GB.utf-8` originally mentioned

When time allows I will whittle away the cruft and close the issue with what I found.  Hopefully this will help others too.  Many thanks.

**Comment by ZyX-I** [comment]
@VoltairesPuppet `powerline-render` is not supposed to use any daemons: it is a fallback when daemon is not found. And I should say that none of built-in scripts run daemon on their own: sometimes it is not needed, some people may prefer to create a systemd service instead, etc. The `default_top_theme: powerline` is exactly a thing which will force unicode theme in non-unicode environment.

If there is a daemon, it is normally (i.e. when you do not pass explicity `--socket` argument) are per-user.

Still did you run second of the suggested commands?

**Comment by VoltairesPuppet** [comment]
### TLDR;
@ZyX-I  I can confirm that I have been able to explicitly reproduce the issue and its definitely down to the daemon capturing the output for some reason.   The detailed steps include use of your python fragment,  which correctly determines `ascii` or `UTF-8` depending on the changing locale - independent of the daemon, 

_p.s. Do you want this to add or your solution to stack-overflow or are you OK if I add it?_

### Detail
I used the python fragment you provided in the steps to confirm the core encoding logic is working: 

	python -c 'from powerline.lib.encoding import get_preferred_output_encoding as gpoe; print(gpoe())'

Steps for reproducing:

- Environment is POSIX for language (originally the default)
	- `export LC_ALL=POSIX`

- Started daemon trying to resolve issues (at which point it presumably captures current locale and determines from that point on the output should be `ascii`):
	- `powerline-daemon`

- Start powerline prompt: 
	- `POWERLINE_BASH_CONTINUATION=1 POWERLINE_BASH_SELECT=1 
           source /usr/lib/python2.7/site-packages/powerline/bindings/bash/powerline.sh`

- Displayed prompt is ascii or question marks if powerline is the forced default.

- Execute python fragment (`python -c 'from powerline.lib.encoding ...`):
	- Output: `ascii`

- Correct language (no change to prompt):
	- `export LC_ALL=en_GB.utf8`

- Execute python fragment:
	- Output: `UTF-8`

- Kill powerline-daemon
	- `kill -9 $(pgrep powerline-daemon)`

- **Resolved!** Prompt immediately switches to whatever the default is.

**Comment by VoltairesPuppet** [comment]
I  have summarised the solution on stackoverflow and "answered" my own question.  As this has resolved my issue, I will mark as closed.  Thanks again @ZyX-I
